### ISSUE TITLE
nn.Parameter init position

### ISSUE THREAD

**Issue by nehSgnaiL** [pytorch/pytorch#127760]
### 🐛 Describe the bug

Hi,

Recently, I found that the different ways to define the trainable parameter might cause different training results, even under the same seed.

My question is: what is the difference between `way 1` and `way 2`? Here are definitions of `way 1` and `way 2`:

```python
class Way1(nn.Module):
    def __init__():
        super(Way1, self).__init__()
        self.wh = nn.Parameter(torch.zeros((37, 19)), requires_grad=True)
        nn.init.xavier_normal_(self.wh)
```

```python
class Way2(nn.Module):
    def __init__(device):
        super(Way2, self).__init__()
        self.device = device
        self.wh = init_weight(37, 19)

    def init_weight(self, shape):
        _para = nn.Parameter(torch.zeros(shape), requires_grad=True).to(self.device)
        nn.init.xavier_normal_(_para)
        return _para
```

I found that initing the trainable parameters in `way 1` seems to get a smaller loss than that in `way 2`. Is it true that if we create an optimizer with net.parameters() as the first way and call optimizer.step(), the self.wh will be automatically optimized and the second way will not?

I would greatly appreciate any insights that point this out for me. :)

### Versions

```cmd
Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.2.1
[pip3] torch_geometric==2.5.3
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl
[conda] mkl                       2023.1.0         h213fc3f_46344
[conda] mkl-service               2.4.0            py39h5eee18b_1
[conda] mkl_fft                   1.3.8            py39h5eee18b_0
[conda] mkl_random                1.2.4            py39hdb19cb5_0
[conda] numpy                     1.26.4           py39h5f9d8c6_0
[conda] numpy-base                1.26.4           py39hb5e798b_0
[conda] pytorch                   2.2.1           py3.9_cuda12.1_cudnn8.9.2_0    pytorch
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch-geometric           2.5.3                    pypi_0    pypi
[conda] torchtriton               2.2.0                      py39    pytorch
```

**Comment by nehSgnaiL** [comment]
Now that I plotted the weights after the training process, it seems that the weight generated in `way2` did not receive the update.
Here are the plots:
|Way|Fig|
|--|--|
|1|![image](https://github.com/pytorch/pytorch/assets/9622749/95c55116-b237-44e2-8ae2-13423683e508)|
|2|![image](https://github.com/pytorch/pytorch/assets/9622749/b6cefc51-4001-4101-af6e-c3bd7cef7a8f)|

**Comment by nehSgnaiL** [comment]
This is probably caused by the `.to(device)` operation; 

refer to [[“Learnable” parameter does not want to learn](https://discuss.pytorch.org/t/learnable-parameter-does-not-want-to-learn/80190)] and a helpful module `torch.nn.ParameterDict` for containing the trainable parameters [[How to add parameters in module class in pytorch custom model?](https://stackoverflow.com/questions/59234238/how-to-add-parameters-in-module-class-in-pytorch-custom-model)].

So I change 
```_para = nn.Parameter(torch.zeros(shape), requires_grad=True).to(self.device)```
to 
```_para = nn.Parameter(torch.zeros(shape, device=self.device), requires_grad=True)```
and it seems to work fine for me so far.
